diff --git a/ex_stage.sv b/ex_stage.sv
index 99145e6..7300544 100644
--- a/ex_stage.sv
+++ b/ex_stage.sv
@@ -274,6 +274,10 @@ module riscv_ex_stage
 		.check_d_i            ( check_d_i_tag           ),
 		.is_load_i            ( regfile_we_i            ),
 		.exception_o          ( exception_o_tag         )
+		`ifdef DIFT_H
+		,
+		.is_store_i			  ( )
+		`endif
 	);
 `endif
 
diff --git a/mode_tag.sv b/mode_tag.sv
index 028da9e..27f6e5e 100644
--- a/mode_tag.sv
+++ b/mode_tag.sv
@@ -43,8 +43,10 @@ module riscv_mode_tag
     register_set_o               = 1'b0;
     memory_set_o                 = 1'b0;
     is_store_post_o              = 1'b0;
+    `ifdef DIFT_H
     memory_set_hierarchy_o       = 1'b0; 
     hierarchy_tag_o              = 2'b00;
+    `endif
 
     unique case (instr_rdata_i[6:0])
 
@@ -241,15 +243,17 @@ module riscv_mode_tag
           endcase
       end
 
+      `ifdef DIFT
       OPCODE_DIFT_STORE: begin
 
         unique case (instr_rdata_i[13:12])
-
+          `ifdef DIFT_H
           2'b11: begin
             // p.hmem
             memory_set_hierarchy_o = 1'b1;
             hierarchy_tag_o <= thr_i[1:0];
           end
+          `endif
 
           default: begin
             // p.sps{w,h,b} 
@@ -259,6 +263,7 @@ module riscv_mode_tag
         endcase
 
       end
+      `endif
 
       OPCODE_DIFT_IMM : begin
 
diff --git a/tag_check_logic.sv b/tag_check_logic.sv
index 3881307..0c8f610 100644
--- a/tag_check_logic.sv
+++ b/tag_check_logic.sv
@@ -12,6 +12,7 @@
 ////////////////////////////////////////////////////////////////////////////////
 
 import riscv_defines::*;
+`include "riscv_config.sv"
 
 module riscv_tag_check_logic
 (
@@ -24,6 +25,11 @@ module riscv_tag_check_logic
   input  logic  is_load_i,
 
   output logic  exception_o
+
+  `ifdef DIFT_H
+  ,
+  input is_store_i
+  `endif
 );
 
   always_comb
